Upgrade to Terraform Plugin SDK v2

v12 opened the pull request:
Upgrade Terraform SDK to v2 and enable debugging the provider as per [the official migration guidelines](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html)

ddelnano commented on mikrotik/resource_bgp_instance.go:
```diff
@@ -120,10 +128,15 @@ func resourceBgpInstanceRead(d *schema.ResourceData, m interface{}) error {
 	}
 
-	return bgpInstanceToData(bgpInstance, d)
+	err = bgpInstanceToData(bgpInstance, d)
```
I'd prefer to update ToData functions to return a `diag.Diagnostics` type rather than update each create, update and delete function.

We could do something like the following to reduce the amount of return statements that would need updated.

```
func bgpInstanceToData(b *client.BgpInstance, d *schema.ResourceData) diag.Diagnostics {
   values := map[string]interface{}{
      "name": b.Name
     "client_to_client_reflection": b.ClientToClientReflection
     
     [ ... ]
   }
   d.SetId(b.Name)
   for attrKey, attrValue := range values { 
     if err := d.Set(attrKey, attrValue); err != nil {
        return diag.FromErr(err)
      }
   }
```

v12 replied:
Fair enough, it sounds like a better way to do it. I didn't focus much on the actual logic inside the code in order not to bloat this particular PR. I want to put some time into a couple more improvements, but I plan to raise them as separate PRs. 🙂

ddelnano reviewed (changes requested):
I still need to look through these changes more, but posting the feedback I have so far. In addition to the comments below, the test suite got significantly slower

Here is the test suite output on master
```
$ make testacc

PASS
ok      github.com/ddelnano/terraform-provider-mikrotik/mikrotik        28.380s
```

Here is the output on this branch
```
$ make testacc

PASS
PASS    github.com/ddelnano/terraform-provider-mikrotik/mikrotik        247.826s
```

That isn't something we have much control over, but it was something that I noticed since the test suite prior was incredibly fast.

v12 commented on mikrotik/provider_test.go:
```diff
@@ -25,7 +27,7 @@ func testAccPreCheck(t *testing.T) {
 	if v := os.Getenv("MIKROTIK_USER"); v == "" {
 		t.Fatal("The MIKROTIK_USER environment variable must be set")
 	}
-	if v := os.Getenv("MIKROTIK_PASSWORD"); v == "" {
+	if _, exists := os.LookupEnv("MIKROTIK_PASSWORD"); !exists {
 		t.Fatal("The MIKROTIK_PASSWORD environment variable must be set")
 	}
```
The main reason to use a `LookupEnv` is to allow using an account without a password (i.e., the value of the environment variable is an empty string). It checks if `MIKROTIK_PASSWORD` environment variable exists and doesn't throw an exception if it's empty.

Happy to switch it back to what it was for consistency, but it may be worth updating the provider to use `LookupEnv` too.

v12 replied:
I checked the rest of the code, and it looks that the behaviour is consistent. There is a check if any of the environment variables is empty in `client/client.go`, but it does nothing in that case:

https://github.com/ddelnano/terraform-provider-mikrotik/blob/d9648a2829c4c7264195607bffbfcd2afaf94464/client/client.go#L173-L175

There are no other places where it'd be used differently.

v12 commented:
Good spot regarding the test suite execution time, I haven't thought of comparing the time it takes before and after the changes. Need to have a look at what may be causing it but 10x increase looks suspicious.

v12 commented:
@ddelnano I refactored entity conversion functions to return `diag.Diagnostics` and updated the test suite to run in parallel, significantly speeding up the whole test run.

ddelnano commented on mikrotik/internal/test_helpers.go:
```diff
@@ -0,0 +1,40 @@
+var macCounter = 0
+var dnsCounter = 0
+
+func GetNewIpAddr() string {
```
I'd prefer to not use these new functions until an ephemeral test environment is in place. I know you mentioned in discord that your other branch depends on these changes, but I'm hesitant to test any branch against my router that doesn't use very specific IPs, mac addresses, etc.

Could we merge the change to add a Github action with the acceptance tests first?

ddelnano commented on client/bgp_peer_test.go:
```diff
@@ -20,6 +18,17 @@ var nextHopChoice string = "default"
+	if err != nil {
+		t.Fatalf("unable to create BGP instance used for testing: %v", err)
+	}
+	defer func(c *Mikrotik, name string) {
```
Have you verified that this works? I know that all deferred functions should run if this is called (reference [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/testing/testing.go;l=723;drc=refs%2Ftags%2Fgo1.17)), but I think the deferred function might need to come before calling `t.Fatalf`.

ddelnano commented on client/bgp_peer_test.go:
```diff
@@ -52,6 +61,17 @@ func TestAddBgpPeerAndDeleteBgpPeer(t *testing.T) {
+	if err != nil {
+		t.Fatalf("unable to create BGP instance used for testing: %v", err)
+	}
+	defer func(c *Mikrotik, name string) {
```
Same question here.

ddelnano replied:
Right :upside_down_face:. Was a late night when I was looking into this.

ddelnano commented on mikrotik/resource_bgp_instance.go:
```diff
@@ -134,102 +137,77 @@ func resourceBgpInstanceUpdate(d *schema.ResourceData, m interface{}) error {
-		return err
+	var diags diag.Diagnostics
+
+	for key, value := range values {
+		if err := d.Set(key, value); err != nil {
+			diags = append(diags, diag.Errorf("failed to set %s: %v", key, err)...)
+		}
 	}
```
This is so much nicer. We could probably make this even better with a function like so since we have this same loop multiple times.

```
func SetResourceDataAttrs(d *schema.ResourceData, attrs map[string]interface{}) diags diag.Diagnostics {
  var diags diag.Diagnostics
  for key, value := range values {
		if err := d.Set(key, value); err != nil {
			diags = append(diags, diag.Errorf("failed to set %s: %v", key, err)...)
		}
	}
  return diags
}
```

This PR is getting big so I don't think we need to do it now, but we should follow up with that afterwards.

ddelnano commented on mikrotik/provider_test.go:
```diff
@@ -25,7 +39,41 @@ func testAccPreCheck(t *testing.T) {
 	}
 }
+
+func testAccDeleteResource(resource *schema.Resource, d *schema.ResourceData, meta interface{}) error {
+	if resource.DeleteContext != nil {
+		var diags diag.Diagnostics
+
+		diags = resource.DeleteContext(context.Background(), d, meta)
+
+		for i := range diags {
+			if diags[i].Severity == diag.Error {
+				return fmt.Errorf("error deleting resource: %s", diags[i].Summary)
+			}
+		}
+
+		return nil
+	}
+
+	return resource.Delete(d, meta)
+}
+
+func testAccCheckResourceDisappears(provider *schema.Provider, resource *schema.Resource, resourceName string) resource.TestCheckFunc {
```
Are these functions necessary for this change? This refactor is long overdue, but there is still many resource tests that aren't using it. Can we add this function in this PR, but defer using it until a future one given this PR's size?

ddelnano reviewed (changes requested):
Apologies for the long review and I greatly appreciate all the effort you've put in. Having CI that runs automatically, upgrading the SDK and all the improvements you've made are amazing.

I've gotten through most of it, but I will try to complete it tomorrow.

If merging the GitHub actions change first isn't an easy option, can you please paste the full output of `make testclient` and `make testacc` for each new commit so I can verify that the tests are all working? The new helper functions that generate random IPs, etc are obviously necessary for this change, but I want to avoid running this against my router given its more dynamic behavior.

ddelnano commented:
Ok, was able to get through the rest of this PR. No new comments from my last review.

ddelnano commented:
This looks awesome, really appreciate all of the hard work into this PR!
